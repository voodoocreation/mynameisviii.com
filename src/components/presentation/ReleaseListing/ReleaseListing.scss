@import "../../../scss/variables", "../../../scss/mixins";

.ReleaseListing {
  flex-basis: 50%;

  a {
    display: block;
    position: relative;
    margin: $spacing-half;
    overflow: hidden;
  }

  h3 {
    margin: 0;
  }

  &--image .Loader {
    padding-top: 100%;
  }

  &--details {
    @include panel-1();

    display: flex;
    flex-direction: column;
    align-items: center;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    padding: $spacing;
    overflow: hidden;
  }

  &--header {
    display: flex;
    align-items: flex-end;
    flex-grow: 1;
    text-align: center;
    padding-bottom: $spacing;
  }

  &--meta {
    flex-grow: 1;
    font-size: 14px;
  }

  @include when-below($screen-xs-max) {
    &--details {
      @include hide-visually();
    }
  }

  @include when-above($screen-xs-min) {
    flex-basis: 33.333333%;

    &--details {
      transform: translate3d(0, 100%, 0);
      transition: transform 200ms ease-out;
    }

    a:hover &--details {
      transform: translate3d(0, 0, 0);
    }
  }

  @include when-above($screen-sm-min) {
    flex-basis: 50%;
  }

  @include when-above(840px) {
    flex-basis: 33.333333%;
  }

  @include when-above(1100px) {
    flex-basis: 25%;
  }

  @include when-above(1320px) {
    flex-basis: 20%;
  }

  @include when-above(1400px) {
    &--meta {
      font-size: 16px;
    }
  }
}
