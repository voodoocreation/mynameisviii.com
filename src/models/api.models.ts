import { arrayToAssoc } from "../transformers/transformData";

interface IWithSlug {
  slug: string;
}

export type TLastEvaluatedKey<K extends string> = Record<
  K | "isActive" | "slug",
  string
>;

export interface IRawDynamoResponse<T extends {}, K extends string> {
  items: T[];
  lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IDynamoResponse<T extends IWithSlug, K extends string> {
  readonly items: Record<string, T>;
  readonly lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IRawS3Response<T extends {}> {
  isTruncated?: boolean;
  items: T[];
}

export interface IS3Response<T extends IWithSlug> {
  readonly isTruncated: boolean;
  readonly items: Record<string, T>;
}

export const dynamoResponse = <T extends IWithSlug, K extends string>(
  options: IRawDynamoResponse<T, K>
): IDynamoResponse<T, K> => ({
  items: arrayToAssoc(options.items, "slug"),
  lastEvaluatedKey: options.lastEvaluatedKey
});

export const s3Response = <T extends IWithSlug>(
  options: IRawS3Response<T>
): IS3Response<T> => ({
  isTruncated: !!options.isTruncated,
  items: arrayToAssoc(options.items, "slug")
});
