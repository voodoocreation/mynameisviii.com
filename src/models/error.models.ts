export interface IError {
  readonly message: string;
  readonly status: number;
}

export const error = (options: Partial<IError> = {}): IError => ({
  message: options.message || "An error has occurred",
  status: options.status || 500,
});
