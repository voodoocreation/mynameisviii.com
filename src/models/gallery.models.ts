import dayjs from "dayjs";

export interface IRawGallery {
  description?: string;
  imageUrl?: string;
  images?: Array<Partial<IGalleryImage>>;
  modifiedAt?: string;
  slug?: string;
  title?: string;
}

export interface IGallery {
  readonly description?: string;
  readonly imageUrl: string;
  readonly images?: IGalleryImage[];
  readonly modifiedAt: string;
  readonly slug: string;
  readonly title: string;
}

export interface IGalleryImage {
  readonly imageUrl: string;
  readonly modifiedAt: string;
}

export const galleryImage = (
  options: Partial<IGalleryImage> = {}
): IGalleryImage => ({
  imageUrl: options.imageUrl || "",
  modifiedAt: options.modifiedAt || dayjs().toISOString()
});

export const gallery = (options: IRawGallery = {}): IGallery => ({
  description: options.description,
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(galleryImage) : undefined,
  modifiedAt: options.modifiedAt || dayjs().toISOString(),
  slug: options.slug || "",
  title: options.title || ""
});
