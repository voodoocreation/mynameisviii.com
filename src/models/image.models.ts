export interface IImage {
  readonly imageUrl: string;
  readonly title: string;
}

export const image = (options: Partial<IImage> = {}): IImage => ({
  imageUrl: options.imageUrl || "",
  title: options.title || ""
});
