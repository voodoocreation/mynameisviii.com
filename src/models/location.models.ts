import { TYPE } from "../constants/location.constants";

export interface ILatLng {
  readonly lat: number;
  readonly lng: number;
}

export interface ILocation {
  readonly address?: string;
  readonly latLng?: ILatLng;
  readonly name: string;
  readonly type: TYPE;
  readonly url?: string;
}

export const location = (options: Partial<ILocation> = {}): ILocation => ({
  address: options.address,
  latLng: options.latLng,
  name: options.name || "",
  type: options.type || TYPE.MUSIC_VENUE,
  url: options.url,
});
