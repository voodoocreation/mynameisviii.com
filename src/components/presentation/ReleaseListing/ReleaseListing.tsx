import cn from "classnames";
import * as React from "react";
import {
  MdAccessTime,
  MdDateRange,
  MdFormatListNumbered,
  MdMusicNote,
} from "react-icons/md";
import {
  FormattedMessage,
  injectIntl,
  WrappedComponentProps,
} from "react-intl";

import { IRelease } from "../../../models/root.models";
import Schema from "../../schema/Release";
import DateTime from "../DateTime/DateTime";
import Image from "../Image/Image";
import Link from "../Link/Link";
import Meta from "../Meta/Meta";

import "./ReleaseListing.scss";

interface IProps extends IRelease, WrappedComponentProps {
  onLoad?: () => void;
}

interface IState {
  isRendered: boolean;
}

class ReleaseListing extends React.Component<IProps, IState> {
  public state: IState = {
    isRendered: false,
  };

  public render() {
    const { intl, onLoad, ...release } = this.props;
    const { isRendered } = this.state;
    const trackCount = release.tracklist.reduce(
      (acc, curr) => acc + curr.length,
      0
    );

    return (
      <article className={cn("ReleaseListing", { isRendered })}>
        <Link route={`/releases/${release.slug}`}>
          <div className="ReleaseListing--details">
            <header className="ReleaseListing--header">
              <h3>
                <span>{release.title}</span>
              </h3>
            </header>

            <section className="ReleaseListing--meta">
              <Meta
                className="ReleaseListing--releasedOn"
                icon={<MdDateRange />}
                labelIntlId="RELEASED"
              >
                <DateTime
                  isRelative={false}
                  options={{
                    day: "numeric",
                    month: "long",
                    year: "numeric",
                  }}
                  value={release.releasedOn}
                  isDateOnly
                />
              </Meta>

              <Meta
                className="ReleaseListing--length"
                icon={<MdAccessTime />}
                labelIntlId="LENGTH"
              >
                {release.length}
              </Meta>

              <Meta
                className="ReleaseListing--genre"
                icon={<MdMusicNote />}
                labelIntlId="GENRE"
              >
                {release.genre}
              </Meta>

              <Meta
                className="ReleaseListing--tracks"
                icon={<MdFormatListNumbered />}
              >
                <FormattedMessage id="X_TRACKS" values={{ trackCount }} />
              </Meta>
            </section>
          </div>

          <Image
            alt={release.title}
            className="ReleaseListing--image"
            src={release.images[0].imageUrl}
            onLoad={this.onLoad}
          />
        </Link>

        <Schema {...release} />
      </article>
    );
  }

  private onLoad = () => {
    this.setState({
      isRendered: true,
    });

    if (this.props.onLoad) {
      this.props.onLoad();
    }
  };
}

export default injectIntl(ReleaseListing);
