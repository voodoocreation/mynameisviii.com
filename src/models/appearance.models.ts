import dayjs from "dayjs";

import { BOOLEAN } from "../constants/api.constants";
import { STATUS, TYPE } from "../constants/appearance.constants";
import { IImage, image } from "./image.models";
import { ILocation, location } from "./location.models";
import { IOffer, offer } from "./offer.models";
import { IOrganization, organization } from "./organization.models";
import { IPerformer, IRawPerformer, performer } from "./performer.models";

export interface IRawAppearance {
  acts?: Array<Partial<IRawPerformer>>;
  audience?: string;
  description?: string;
  finishingAt?: string;
  imageUrl?: string;
  images?: IImage[];
  isActive?: BOOLEAN;
  location?: Partial<ILocation>;
  ogImageUrl?: string;
  organizer?: Partial<IOrganization>;
  rsvpUrl?: string;
  sales?: Array<Partial<IOffer>>;
  slug?: string;
  startingAt?: string;
  status?: STATUS;
  title?: string;
  type?: TYPE;
}

export interface IAppearance {
  readonly acts: IPerformer[];
  readonly audience?: string;
  readonly description: string;
  readonly finishingAt: string;
  readonly imageUrl: string;
  readonly images: IImage[];
  readonly isActive: boolean;
  readonly location: ILocation;
  readonly ogImageUrl: string;
  readonly organizer: IOrganization;
  readonly rsvpUrl?: string;
  readonly sales: IOffer[];
  readonly slug: string;
  readonly startingAt: string;
  readonly status: STATUS;
  readonly title: string;
  readonly type: TYPE;
}

export const appearance = (options: IRawAppearance = {}): IAppearance => ({
  acts: options.acts ? options.acts.map(performer) : [],
  audience: options.audience,
  description: options.description || "",
  finishingAt: options.finishingAt || dayjs().toISOString(),
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(image) : [],
  isActive: (options.isActive && options.isActive === BOOLEAN.TRUE) || false,
  location: location(options.location),
  ogImageUrl: options.ogImageUrl || "",
  organizer: organization(options.organizer),
  rsvpUrl: options.rsvpUrl,
  sales: options.sales ? options.sales.map(offer) : [],
  slug: options.slug || "",
  startingAt: options.startingAt || dayjs().toISOString(),
  status: options.status || STATUS.SCHEDULED,
  title: options.title || "",
  type: options.type || TYPE.MUSIC,
});
