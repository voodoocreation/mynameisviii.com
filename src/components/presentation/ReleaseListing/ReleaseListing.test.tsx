import { release } from "../../../models/root.models";
import WrapperWithIntl from "../../../utilities/WrapperWithIntl";
import ReleaseListing from "./ReleaseListing";

const component = new WrapperWithIntl(ReleaseListing).withDefaultProps(
  release({
    artist: {
      name: "Artist Name",
      url: "URL"
    },
    buyList: [],
    description: "Description",
    genre: "Genre",
    images: [
      {
        imageUrl: "Image URL",
        title: "Image title"
      }
    ],
    length: "5:00",
    recordLabel: "Record Label",
    releasedOn: "2017-01-01T00:00:00",
    slug: "release-1",
    streamList: [],
    title: "Title",
    tracklist: [
      [
        {
          genre: "Genre",
          length: "5:00",
          title: "Title",
          url: "URL"
        }
      ]
    ]
  })
);

describe("[presentation] <ReleaseListing />", () => {
  const wrapper = component
    .withProps({
      onLoad: jest.fn()
    })
    .mount();

  it("doesn't add the isRendered class initially", () => {
    expect(wrapper.find(".isRendered")).toHaveLength(0);
  });

  it("loads the image", () => {
    wrapper.find("Image.ReleaseListing--image").simulate("load");
  });

  it("adds the isRendered class", () => {
    expect(wrapper.find(".isRendered")).toHaveLength(1);
  });

  it("calls the onLoad prop", () => {
    expect(component.props.onLoad).toHaveBeenCalledTimes(1);
  });

  it("sets the onLoad prop to be undefined", () => {
    wrapper.setProps({ onLoad: undefined });
  });

  it("loads the image", () => {
    wrapper.find("Image.ReleaseListing--image").simulate("load");
  });

  it("matches snapshot", () => {
    expect(wrapper.render()).toMatchSnapshot();
  });
});
