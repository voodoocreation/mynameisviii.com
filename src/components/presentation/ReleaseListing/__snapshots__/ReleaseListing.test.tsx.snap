// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <ReleaseListing /> matches snapshot 1`] = `
<article
  class="ReleaseListing isRendered"
>
  <a
    href="/releases/release-1"
  >
    <div
      class="ReleaseListing--details"
    >
      <header
        class="ReleaseListing--header"
      >
        <h3>
          <span>
            Title
          </span>
        </h3>
      </header>
      <section
        class="ReleaseListing--meta"
      >
        <div
          class="Meta ReleaseListing--releasedOn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 11H7v2h2v-2zm4 0h-2v2h2v-2zm4 0h-2v2h2v-2zm2-7h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V9h14v11z"
            />
          </svg>
           
          <span
            class="Meta--label"
          >
            Released:
          </span>
           
          <span
            class="Meta--value"
          >
            <time
              class="DateTime"
              datetime="2017-01-01"
              title="1 January 2017"
            >
              1 January 2017
            </time>
          </span>
        </div>
        <div
          class="Meta ReleaseListing--length"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm.5-13H11v6l5.25 3.15.75-1.23-4.5-2.67z"
            />
          </svg>
           
          <span
            class="Meta--label"
          >
            Length:
          </span>
           
          <span
            class="Meta--value"
          >
            5:00
          </span>
        </div>
        <div
          class="Meta ReleaseListing--genre"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 3v10.55c-.59-.34-1.27-.55-2-.55-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4V7h4V3h-6z"
            />
          </svg>
           
          <span
            class="Meta--label"
          >
            Genre:
          </span>
           
          <span
            class="Meta--value"
          >
            Genre
          </span>
        </div>
        <div
          class="Meta ReleaseListing--tracks"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2 17h2v.5H3v1h1v.5H2v1h3v-4H2v1zm1-9h1V4H2v1h1v3zm-1 3h1.8L2 13.1v.9h3v-1H3.2L5 10.9V10H2v1zm5-6v2h14V5H7zm0 14h14v-2H7v2zm0-6h14v-2H7v2z"
            />
          </svg>
            
          <span
            class="Meta--value"
          >
            1 track
          </span>
        </div>
      </section>
    </div>
    <figure
      class="Image ReleaseListing--image isLoading"
    >
      <div
        class="Image--wrapper"
      >
        <img
          alt="Title"
          src="Image URL"
        />
        <span
          class="Loader"
        >
          <div
            class="Loader--animation"
          >
            <span />
            <span />
            <span />
          </div>
        </span>
      </div>
    </figure>
  </a>
  <script
    type="application/ld+json"
  >
    {"@id":"http://localhost:/releases/release-1","@type":"MusicAlbum","albumProductionType":"http://schema.org/StudioAlbum","albumRelease":{"@type":"MusicRelease","duration":"PT5M0S","name":"Title","recordLabel":"Record Label"},"albumReleaseType":"http://schema.org/album","byArtist":{"@type":"MusicGroup","name":"Artist Name"},"datePublished":"2017-01-01T00:00:00","description":"Description","image":"Image URL","mainEntityOfPage":{"@id":"http://localhost:/releases/release-1"},"name":"Title","track":{"@type":"ItemList","itemListElement":[{"@type":"ListItem","item":{"@type":"MusicRecording","duration":"PT5M0S","name":"Title","recordingOf":{"@type":"MusicComposition","name":"Title"}},"position":1}],"numberOfItems":1},"@context":"http://schema.org/"}
  </script>
</article>
`;
